Added Menu Validity Event

HenryLoenwind opened the pull request:
An event to override the distance checks that close menus when the player is too far away.

This can be used by mods to provide a means of remote access, like the old "Ender IO" of Ender IO.

The original "CanStayOpen" event got lost with the big license change and a replacement failed to get through for the same reasons neo forked. This code is not a copy of the old event, neither in name or code, so it is not tainted by its license.

Note that I added this in a minimally invasive way, so the event hasn't much context and it is not enforced that modded menus also fire it instead of doing the distance check themselves. However, as the currently open menu is easily accessible and there's no good way to force a modded menu to stay open if it doesn't want to, this is, in my opinion, sufficient.

HenryLoenwind committed:
Added Menu Validity Event

Spinoscythe commented on src/main/java/net/neoforged/neoforge/event/ValidityEvent.java:
```diff
@@ -0,0 +1,104 @@
+        return result.orElse(originalValue);
+    }
+
+    public static class Menu extends ValidityEvent {
```
My idea is make this the default impl and let `ContainerMenu` and `Sign` be the only nested impls. Also, rename the top level event to `MenuValidityEvent`

HenryLoenwind replied:
I had it this way around at first, but that lacked a way of indicating that there is no menu open. So any mod trusting that a "*Menu*ValidityEvent" would only be fired when there's a menu set for the player would work fine...until someone opened a sign's ~~menu~~ *screen* and it'd crash horribly. That's why I moved the Menu variant down to be a sibling of Sign.

Spinoscythe replied:
You know what, you're right. But I would personally say that `ValidityEvent` isn't a good name. Potential rename?

HenryLoenwind replied:
I'm not happy with that name either and hoped to get a good suggestion here ;).

"CanStayOpenEvent" would be better, but that's too close to the old wrong-licence one, and I want to avoid any suspicion of copying. Also, the method this patches in is now called stillValid(), not canStayOpen().

Spinoscythe replied:
`PlayerRemoteAccessEvent`?

HenryLoenwind replied:
How about MenuRangeCheckEvent? As it only replaces the range check part of the validity check...

Spinoscythe replied:
@HenryLoenwind `MenuDistanceValidationEvent`?

RaymondBlaze replied:
> I had it this way around at first, but that lacked a way of indicating that there is no menu open. So any mod trusting that a "MenuValidityEvent" would only be fired when there's a menu set for the player would work fine...until someone opened a sign's ~menu~ _screen_ and it'd crash horribly. That's why I moved the Menu variant down to be a sibling of Sign.

Just don't let the subscriber assert anything they don't know from the event, if they want to assert anything they should do that themselves.
The child event should only provide basic information that is not open to the subscriber's scope.
The `Sign` child event is pretty much pointless as the `SignBlockEntity` can be retrieved by `Level#getBlockEntity` anyways.
Do note that modded blocks may also use a menuless screen, I don't believe they should be forced to create their own variant of the event like you did for signs.
The same applies to the `ContainerEntity` child event, if you simply don't inline the original check in the constructor, there's no point separating it from the `Entity` child event.
I believe there should only be a `Block` child event providing only the `BlockPos` context, and a `Entity` child event providing only the `Entity` context, the subscriber should retrieve specific context themselves and only make assertions based on their own checks (like `instanceof ContainerEntity`).
Considering entity also has a valid `BlockPos`, and a position is needed anyways as the distance is in the base event, the `Block` child can optionally be pulled up to the top level, and let the `Entity` child be a separated entity specific implementation.

HenryLoenwind replied:
> > I had it this way around at first, but that lacked a way of indicating that there is no menu open. So any mod trusting that a "MenuValidityEvent" would only be fired when there's a menu set for the player would work fine...until someone opened a sign's ~menu~ _screen_ and it'd crash horribly. That's why I moved the Menu variant down to be a sibling of Sign.
> 
> Just don't let the subscriber assert anything they don't know from the event, if they want to assert anything they should do that themselves. The child event should only provide basic information that is not open to the subscriber's scope.

I can't prevent modders from assuming a menu is open when they subscribe to a "menu" event. That's why the event provides the information "hey this is not for a menu". When firing the event, the code knows that and can pass it on.

> The `Sign` child event is pretty much pointless as the `SignBlockEntity` can be retrieved by `Level#getBlockEntity` anyways.

But how would a subscriber know that? They could guess based on what's at the blockpos, but I prefer telling them explicitly. Also, when I already have the BE on hand, why have the subscriber look it up again? That just wastes CPU cycles and handing it over in the event is as close to free as possible.

> Do note that modded blocks may also use a menuless screen, I don't believe they should be forced to create their own variant of the event like you did for signs.

And that is a valid point. Changing the "Sign" sub-event to "Screen" and the parameter to a BE certainly makes sense.

This leaves Screens that are bound to a location that has no BE and no menu. Those don't exist in vanilla, and so I'd be reluctant to add a specific event for those. Too much guesswork about how that would be used for my liking, although I could make the "Blockbased" constructor public...

> The same applies to the `ContainerEntity` child event, if you simply don't inline the original check in the constructor, there's no point separating it from the `Entity` child event. I believe there should only be a `Block` child event providing only the `BlockPos` context, and a `Entity` child event providing only the `Entity` context,

The ContainerEntity event exists because the container that fires the event does not have access to the Entity (and vice versa). All I could provide is an AABB, which is mostly useless for finding the original Entity. For the same reason, the "Menu" event doesn't provide the menu/BE/etc. It is not available where it is fired, so the event subscriber's guess is as good as mine.

> the subscriber should retrieve specific context themselves and only make assertions based on their own checks (like `instanceof ContainerEntity`). Considering entity also has a valid `BlockPos`, and a position is needed anyways as the distance is in the base event,

The Entity has a pos, the ContainerEntity doesn't. That's why I removed Level and BlockPos from the base event. I could fill them in the Entity case (as I originally had it), but after pulling it down, that wasn't needed anymore. There's no advantage to putting stuff from entity in 2 additional fields when the entity itself is available.

> the `Block` child can optionally be pulled up to the top level, and let the `Entity` child be a separated entity specific implementation.

Making the top-level event abstract and opening up the Blockbased one instead is something I'd be fine with. Block-based and entity-based have diverged enough to make subscribing to the top-level event mostly pointless.

HenryLoenwind replied:
PS: `ContainerEntity` extends `Container`, NOT `Entity`. Its name is wardsback.

Spinoscythe commented on src/main/java/net/neoforged/neoforge/event/ValidityEvent.java:
```diff
@@ -0,0 +1,104 @@
+    public static class Sign extends ValidityEvent {
+        private final SignBlockEntity signBlockEntity;
+
+        public Sign(Player player, Level level, BlockPos blockPos, double distance, SignBlockEntity signBlockEntity) {
```
Mark this constructor as `@ApiStatus.Internal`

Spinoscythe commented on src/main/java/net/neoforged/neoforge/event/ValidityEvent.java:
```diff
@@ -0,0 +1,104 @@
+    public static class ContainerMenu extends Menu {
+        private final AbstractContainerMenu menu;
+
+        public ContainerMenu(Player player, Level level, BlockPos blockPos, double distance, AbstractContainerMenu menu) {
```
Mark this constructor as `@ApiStatus.Internal`

HenryLoenwind committed:
@ApiStatus.Internal

HenryLoenwind committed:
Formatted differently than the checked-in format file says

It clearly states there should be an empty line between import groups...

HenryLoenwind committed:
Merge remote-tracking branch 'neoforged/1.21.x' into open

HenryLoenwind committed:
Added entity-based menus

and removed that one event subtype that was fired from one single container only.

RaymondBlaze commented on src/main/java/net/neoforged/neoforge/event/ValidityEvent.java:
```diff
@@ -0,0 +1,133 @@
+ */
+public class ValidityEvent extends PlayerEvent {
+    private final double distance;
+    private final boolean originalValue;
+    private Optional<Boolean> result = Optional.empty();
```
```suggestion
    private final boolean isOriginallyValid;
    private boolean isValid;
```
No need to use `Optional` for field with setters, and `originalValue`/`result` is simply not self-explanatory, especially when you didn't add javadocs to the methods (yeah, please add javadoc for the public methods).

RaymondBlaze commented on src/main/java/net/neoforged/neoforge/event/ValidityEvent.java:
```diff
@@ -0,0 +1,133 @@
+        return distance;
+    }
+
+    public boolean getOriginalValue() {
+        return originalValue;
```
```suggestion
    public boolean isOriginallyValid() {
        return isOriginallyValid;
```
Match the suggested change on field names above.

RaymondBlaze commented on src/main/java/net/neoforged/neoforge/event/ValidityEvent.java:
```diff
@@ -0,0 +1,133 @@
+        return originalValue;
+    }
+
+    public void setResult(boolean result) {
+        this.result = Optional.of(result);
```
```suggestion
    public void setIsValid(boolean isValid) {
        this.isValid = isValid;
```

RaymondBlaze commented on src/main/java/net/neoforged/neoforge/event/ValidityEvent.java:
```diff
@@ -0,0 +1,133 @@
+        this.result = Optional.of(result);
+    }
+
+    public boolean getResult() {
+        return result.orElse(originalValue);
```
```suggestion
    public boolean isValid() {
        return isValid;
```
Match the suggested change on field names above.

RaymondBlaze commented on src/main/java/net/neoforged/neoforge/event/ValidityEvent.java:
```diff
@@ -0,0 +1,133 @@
+ */
+public class ValidityEvent extends PlayerEvent {
+    private final double distance;
+    private final boolean originalValue;
+    private Optional<Boolean> result = Optional.empty();
```
```suggestion
    private final boolean isOriginallyValid;
    private boolean isValid;
```
No need to use `Optional` for field with setter, the field name should be more self-explanatory.

HenryLoenwind replied:
I used the Optional for its nice orElse(), which makes the code that produces the final result more expressive. However, looking at the event again, it seems like I forgot separate getters for the original/override. This means a subscriber cannot distinguish between the original value and something another earlier subscriber set. Oops.

result->isValid: Yes, that naming was habit and doesn't match how neo does events. I'll fix that.

RaymondBlaze commented on src/main/java/net/neoforged/neoforge/event/ValidityEvent.java:
```diff
@@ -0,0 +1,133 @@
+    private final boolean originalValue;
+    private Optional<Boolean> result = Optional.empty();
+
+    ValidityEvent(Player player, double distance, boolean originalValue) {
+        super(player);
+        this.distance = distance;
+        this.originalValue = originalValue;
+    }
```
```suggestion
    ValidityEvent(Player player, double distance, boolean isValid) {
        super(player);
        this.distance = distance;
        this.isValid = this.isOriginallyValid = isValid;
    }
```
Match the change to the field.

RaymondBlaze commented on src/main/java/net/neoforged/neoforge/event/ValidityEvent.java:
```diff
@@ -0,0 +1,133 @@
+        return distance;
+    }
+
+    public boolean getOriginalValue() {
+        return originalValue;
+    }
```
```suggestion
    public boolean isOriginallyValid() {
        return isOriginallyValid;
    }
```
Match the changes to the field.

RaymondBlaze commented on src/main/java/net/neoforged/neoforge/event/ValidityEvent.java:
```diff
@@ -0,0 +1,133 @@
+        return originalValue;
+    }
+
+    public void setResult(boolean result) {
+        this.result = Optional.of(result);
+    }
```
```suggestion
    public void setIsValid(boolean isValid) {
        this.isValid = isValid;
    }
```
Match the changes to the field.

RaymondBlaze commented on src/main/java/net/neoforged/neoforge/event/ValidityEvent.java:
```diff
@@ -0,0 +1,133 @@
+        this.result = Optional.of(result);
+    }
+
+    public boolean getResult() {
+        return result.orElse(originalValue);
+    }
```
```suggestion
    public boolean isValid() {
        return isValid;
    }
```
Match the changes to the field.

HenryLoenwind replied:
You don't need to comment on every single use of that field, I know how to rename a field and cascade the change.

RaymondBlaze commented on src/main/java/net/neoforged/neoforge/common/CommonHooks.java:
```diff
@@ -1558,4 +1561,20 @@ public static Map<RecipeBookType, Pair<String, String>> buildRecipeBookTypeTagFi
         return vanillaMap;
     }
+
+    public static boolean menuValidity(Player player, Level level, BlockPos blockPos, double distance) {
+        return NeoForge.EVENT_BUS.post(new ValidityEvent.Menu(player, level, blockPos, distance)).getResult();
+    }
+
+    public static boolean menuValidity(Player player, Entity entity, double distance) {
+        return NeoForge.EVENT_BUS.post(new ValidityEvent.EntityMenu(player, entity, distance)).getResult();
+    }
+
+    public static boolean menuValidity(Player player, ContainerEntity entity, double distance) {
+        return NeoForge.EVENT_BUS.post(new ValidityEvent.ContainerEntityMenu(player, entity, distance)).getResult();
+    }
+
+    public static boolean menuValidity(Player player, Level level, BlockPos blockPos, double distance, SignBlockEntity signBlockEntity) {
+        return NeoForge.EVENT_BUS.post(new ValidityEvent.Sign(player, level, blockPos, distance, signBlockEntity)).getResult();
+    }
```
I believe these should goes to the `IPlayerExtension` instead, and mark the original methods it replaced as `@Deprecated`.
Otherwise, the original checking method should not be inlined in the event constructor, just leave it in the original call site for not breaking specific mixins from old implementation and multiloader mods.

HenryLoenwind replied:
> I believe these should goes to the `IPlayerExtension` instead, and mark the original methods it replaced as `@Deprecated`.

That doesn't work, as there are other callers to those methods. If I replaced those, I could just as well have added an "AdjustPlayerReachDistanceEvent"...

RaymondBlaze commented on src/main/java/net/neoforged/neoforge/event/ValidityEvent.java:
```diff
@@ -0,0 +1,133 @@
+ * </ul>
+ * This event is fired on the {@link NeoForge#EVENT_BUS}.
+ */
+public class ValidityEvent extends PlayerEvent {
```
The event children is pretty complicated due to inlining the original checking method in the event constructor.
Since we only holds the original boolean result anyways, just pass in the boolean original, it's the subscriber's choice to decide how they would modify the boolean, and they should be aware of the original check if they are only subscribing for certain entities.

Fuzss commented:
I'm not convinced patching all the `stillValid` methods is the right approach for this. 

It's very likely for modders to forget to call the NeoForge `menuValidity` method instead of vanilla. This makes the event a lot less useful for long distance container access implementations, as modded menus are not guaranteed to work. The event shouldn't be optional for modded menus.
Also supporting the event on modded menus in multi-loader projects is gonna be an annoyance probably not worth the hassle.

I think patching the calls to `stillValid` in `ServerPlayer` and `ServerGamePacketListenerImpl` would be more ideal. The event would provide even less context then (only the menu, or block entity in case of signs), but that should be enough, since the vanilla checks can easily be replicated depending on the situation.

Shadows-of-Fire commented:
I agree that patching individual menus is unlikely to be a good long-term solution.  Honestly, having just the menu buys you effectively all the context received with the invasive patches, since that data is coming from the menu in the first place.

HenryLoenwind commented:
> It's very likely for modders to forget to call the NeoForge `menuValidity` method instead of vanilla. This makes the event a lot less useful for long distance container access implementations, as modded menus are not guaranteed to work. The event shouldn't be optional for modded menus. Also supporting the event on modded menus in multi-loader projects is gonna be an annoyance probably not worth the hassle.

Modded menus do NOT have to call Neo code. They have to call what in any sane implementation would be `super.isValid()`, but in Mojang's infinite wisdom, is a call to a static method in the superclass. If they don't do this, their menus will not close if the player moves too far away anyway. And if they check the distance manually, there's no way of preventing this anyway. We cannot take away the agency of a menu to close itself.

> I think patching the calls to `stillValid` in `ServerPlayer` and `ServerGamePacketListenerImpl` would be more ideal. The event would provide even less context then (only the menu, or block entity in case of signs), but that should be enough, since the vanilla checks can easily be replicated depending on the situation.

This would majorly break compatibility in multi-loader setups as we would have to patch out all distance checks in vanilla code and add a new one at the caller level, completely changing how the system works. (It would also prevent menus from having shorter or longer valid distances unless they also subscribe to the event for themselves.) And it would introduce breakage, as menus expecting to be responsible for the distance check now don't have to do it anymore but only in Neo. So, a menu that works fine with Neo wouldn't work with other loaders anymore.

HenryLoenwind commented:
Just for reference, as it's not very obvious, this is the call hierarchy to the distance checks:

Take note that on the block-based side, all but one checks are in one of the static methods (the red S on the icon) in one of the super classes. The Entity side is a bit more sane, but still split into two different branches.

![image](https://github.com/user-attachments/assets/e28eee5b-b52b-4ead-a539-69d415e90ba5)
![image](https://github.com/user-attachments/assets/64a287a0-afff-4f7c-85d4-bca86afcc833)

HenryLoenwind committed:
Cleanup incorporating suggestions

HenryLoenwind committed:
Merge remote-tracking branch 'neoforged/1.21.x' into open

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/event/ValidityEvent.java:
```diff
@@ -0,0 +1,134 @@
+
+    public static class Menu extends BlockBased {
+        @ApiStatus.Internal
+        public Menu(Player player, Level level, BlockPos blockPos, double distance, boolean originallyValid) {
```
I'm not so sure on this being internal. If you're expecting the `stillValid` check to end up firing the event, a mod may have a container whose validity is not based on a distance, but on some other arbitrary value, meaning they'd need to fire the event themselves, and the `CommonHooks` method won't help them.

HenryLoenwind replied:
Because the whole Event tree is only about the distance (I think I'll rename it. I've been avoiding the old name because of the license, but the new one just isn't good.), so its result should not overrule any non-distance-based checks. For this, there are 3 different things to calculate the distance to, BlockPos, AABB, and Entity. The first case has 2 sub-cases, one in which a menu is open and the hacky way signs use pure Screens to mimic what menus do. (A remnant from the time when menus required the block to have an inventory.)

If an event subscriber only cares about the BlockPos and distance, they can subscribe to BlockBased. If a modded block wants to have a GUI, they better have a menu and not copy the whacky thing Sign does, so they can fire Menu.

All constructors are internal/package because there are entry points in CommonHooks for all cases.

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/event/ValidityEvent.java:
```diff
@@ -0,0 +1,134 @@
+        return stillValid.orElse(originallyValid);
+    }
+
+    public static class BlockBased extends ValidityEvent {
```
Why is this not abstract? I'd also prefer if you made the subclasses sealed instead of package-private ctors.

HenryLoenwind replied:
see above

HenryLoenwind committed:
Merge branch '1.21.x' into open

HenryLoenwind committed:
Changed naming and expanded javadoc to make use case clearer

Also opened for special modded subclasses, in case some mod has a server-based GUI that is neither a menu, nor has a BE or Entity.

HenryLoenwind commented:
I renamed the event to make it clearer what it is there for and what it isn't. I also expanded the Javadoc to include answers to questions asked here, as I expect modders looking at this to have the same reaction.

Again, modded menus will very rarely have to manually fire his event. The typical modded menu will call its vanilla superclass's check and that has the event patched in. Only menus that completely ignore their superclass and do everything manually need to fire the event manually. This is the same as with any mod-provided subclass of any vanilla class that has events patched in. If you don't call super, you have to fire events that super would fire manually. If you do call super, that event is fired automatically. The only difference here is that the super call goes to a static method because Mojang needed mixed inheritance and didn't use an interface for that but made two inheritance trees that cross-call each other's static methods.

Matyrobbrt commented:
Given the lack of activity and the need for a rebase, I'll be closing this PR for now. Feel free to open it again after the conflicts are resolved.
